src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs

```diff
@@ -239,6 +262,44 @@ private void ExecuteRemove(ProjectItemGroupTaskItemInstance child, ItemBucket bu
             }
         }
 
+        private List<ProjectItemInstance> FindItemsUsingMatchOnMetadata(
+            ICollection<ProjectItemInstance> items,
+            ProjectItemGroupTaskItemInstance child,
+            ItemBucket bucket,
+            HashSet<string> matchOnMetadata)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(matchOnMetadata, nameof(matchOnMetadata));
+
+            var itemSpec = new ItemSpec<ProjectPropertyInstance, ProjectItemInstance>(child.Remove, bucket.Expander, child.RemoveLocation, Project.Directory, true);
+
+            // todo create OM_MatchOnMetadataIsRestrictedToOnlyOneReferencedItem resource
+            ErrorUtilities.VerifyThrowInvalidOperation(
+                itemSpec.Fragments.Count == 1
+                && itemSpec.Fragments.First() is ItemSpec<ProjectPropertyInstance, ProjectItemInstance>.ItemExpressionFragment,
+                "OM_MatchOnMetadataIsRestrictedToOnlyOneReferencedItem",
+                child.RemoveLocation,
+                child.Remove);
+
+            var itemFragment = itemSpec.Fragments.First() as ItemSpec<ProjectPropertyInstance, ProjectItemInstance>.ItemExpressionFragment;
+
+            var itemsToRemove = new List<ProjectItemInstance>();
+
+            foreach (var item in items)
+            {
+                foreach (var referencedItem in itemFragment.ReferencedItems)
+                {
+                    if (matchOnMetadata.All(
+                            m =>
+                                item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m)) && !item.GetMetadataValue(m).Equals(string.Empty)))
```

Comment:
I know GetMetadataValue isn't supposed to return null, but part of me doesn't like relying on that. String.IsNullOrEmpty?